### DISCUSSION TITLE
Compatibility of kenlm language model between beam search and inference time

### DISCUSSION THREAD

**Discussion by feddybear** [NVIDIA-NeMo/NeMo#4673] [Q&A] [answered]
This is actually more of a Riva-related question but since beam search decoding is part of NeMo, I thought of asking it here.

On the scripts for using kenLM-based LMs for ngram-based rescoring for subword (BPE) models, there is a trick used, which chunks the text encodings and use them for the LM statistics instead of the actual graphemes.

When I try to generate the binary kenLM model based on the beam search scripts, then use this model in flashlight decoder on Riva, no matter what beam settings I use, the inference output comes out as garbage.  I can confirm that in the beam search rescoring, I can get good SotA results on a very challenging dataset in the language that I am working on, so the garbage output doesn't make sense to me.

I was just wondering if flashlight can handle the trick mentioned above, or do I need to remove the trick and use the direct graphemes instead?  This doesn't seem to be clearly stated in any of the documentation.  Kindly help, thanks!

**Comment by titu1994** [comment, marked as answer]
We don't support flashlight, so compatibility with it is not guaranteed. Nemo does do double encoding, so it's probably better to subtract the shift value (default of 100 for Nemo) and then detokenize the subword I'd back to subword text before passing into Flashlight. 

Again, we haven't tried it so no guarantees it would work. But technically, the encoding into Unicode space is easy to reverse (subtract 100 from the token  in char space).

**Reply by titu1994** [reply]
Riva does not apply these tricks, so it would be far easier to just use rivas scripts to build an lm and use that in flashlight as Riva decoder should support flashlight

**Reply by feddybear** [reply]
Awesome..thank you, @titu1994 !
